build-system-abstraction.rst

```diff
@@ -0,0 +1,360 @@
+with the instalation components of easy-install.
+
+As written it depends on PEP-426 which is draft. Sadly PEP-345 does not define setup-requires/build-requires metadata, so we either need a narrow PEP to add that to PEP-345, or to ratify PEP-426, or to break out just-enough-from-pep-426 into a new narrow spec.
+
+Motivation
+==========
+
+There is significant pent-up frustration in the Python packaging ecosystem
+around the current lock-in between build system and pip. Breaking that lock-in
+is better for pip, for setuptools, and for other build systems like flit [#flit]_.
+
+Specification
+=============
+
+Overview
+--------
+
+Build tools will be located by reading a file ``pypa.yaml`` from the root
+directory of the source tree. That file describes how to get the build tool.
+The build tool is then self-describing. This avoids having stale descriptions
+of build tools encoded into sdists that have been uploaded to PyPI [#pypi]_.
+
+The interface involves executing processes rather than loading Python modules,
+because tools like pip will often interact with many different versions of the
+same build tool during a single invocation, and in-process APIs are much harder
+to manage in that situation.
+
+Process interface
+-----------------
+
+Where a process needs to be run, we need to be able to assemble the subprocess
+call to make. For this we use a simple command string with in-Python
+variable interpolation. Only variables defined in this PEP will be
+honoured: the use of additional variables is an error. Additional variables
+can only be added with a schema version increase. Basic variable expressions
+such as ``${PYTHON}``, ``${PYTHON:-python}``, ``${PYTHON:+PYTHON -m coverage}``
+can be used. An implementation of this is in shellvars [#shellvars]_.
+
+Processes will be run with the current working directory set to the root of the source tree.
+
+Available variables
+-------------------
+
+PYTHON
+    The Python interpreter in use.
+OUTPUT_DIR
+    Where to create requested output. If not set,
+    outputting to the current working directory is appropriate.
+
+pypa.yaml
+---------
+
+The yaml file has the following schema. Extra keys are ignored.
+
+schema
+    The version of the schema. This PEP defines version 1.
+    Defaults to 1
+bootstrap-requires
+    The requirements required to run the build tool itself.
+    
+    This should be a list of specifiers, in plain text form.
+    
+    e.g.::
+    
+        bootstrap-requires:
+          - requirement
+          - requirement[extra]
+          - requirement:python_version<"3"
+          - requirement[extra]:python_version>"2"
+          - requirement[extra]>2.5:python_version=="2.7"
+
+build-tool
+    A command to run to query the build tool for its complete interface.
+    The build tool should output on stdout a build tool description JSON document.
+    Stdin may not be read from, and stderr may be handled however the calling process
+    desires.
```

Comment:
Examples:

    flit --dump-build-description
    make self-describe

Not sure if ``make``, ``cMake`` & co meet the stdin/stdout/stderr requirements here. That would be desirable, right?